tests/integration/ceph_upgrade_test.go

```diff
@@ -119,46 +148,99 @@ func (s *UpgradeSuite) TestUpgradeToMaster() {
 	time.Sleep(15 * time.Second)
 
-	err = s.k8sh.WaitForLabeledPodsToRun("app=rook-ceph-mon", s.namespace)
-	require.Nil(s.T(), err)
+	monsNotOldVersion = fmt.Sprintf("app=rook-ceph-mon,rook-version!=%s", oldRookVersion)
+	err = s.k8sh.WaitForDeploymentCount(monsNotOldVersion, s.namespace, numMons)
+	require.NoError(s.T(), err)
+	err = s.k8sh.WaitForLabeledDeploymentsToBeReady(monsNotOldVersion, s.namespace)
+	require.NoError(s.T(), err)
 
 	// wait for the osd pods to be updated
-	osdsNotOldVersion := fmt.Sprintf("app=rook-ceph-osd,rook-version!=%s", oldRookVersion)
+	osdsNotOldVersion = fmt.Sprintf("app=rook-ceph-osd,rook-version!=%s", oldRookVersion)
 	err = s.k8sh.WaitForDeploymentCount(osdsNotOldVersion, s.namespace, numOSDs)
 	require.NoError(s.T(), err)
+	err = s.k8sh.WaitForLabeledDeploymentsToBeReady(osdsNotOldVersion, s.namespace)
+	require.NoError(s.T(), err)
+
+	mdsesNotOldVersion = fmt.Sprintf("app=rook-ceph-mds,rook-version!=%s", oldRookVersion)
+	err = s.k8sh.WaitForDeploymentCount(mdsesNotOldVersion, s.namespace, 4 /* always expect 4 mdses */)
+	require.NoError(s.T(), err)
+	err = s.k8sh.WaitForLabeledDeploymentsToBeReady(mdsesNotOldVersion, s.namespace)
+	require.NoError(s.T(), err)
 
 	logger.Infof("Done with automatic upgrade to master")
 
 	// Give a few seconds for the daemons to settle down after the upgrade
 	time.Sleep(5 * time.Second)
 
+	// wait for filesystem to be active
+	err = waitForFilesystemActive(s.k8sh, s.namespace, filesystemName)
+	require.NoError(s.T(), err)
+
 	// Test writing and reading from the pod with cephfs mounted that was created before the upgrade.
 	postFilename := "post-upgrade-file"
-	assert.Nil(s.T(), s.k8sh.ReadFromPodRetry(s.namespace, filePodName, preFilename, message, 5))
-	assert.Nil(s.T(), s.k8sh.WriteToPodRetry(s.namespace, filePodName, postFilename, message, 5))
-	assert.Nil(s.T(), s.k8sh.ReadFromPodRetry(s.namespace, filePodName, postFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.ReadFromPodRetry(s.namespace, filePodName, preFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.WriteToPodRetry(s.namespace, filePodName, postFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.ReadFromPodRetry(s.namespace, filePodName, postFilename, message, 5))
 
 	// Test writing and reading from the pod with rbd mounted that was created before the upgrade.
 	// There is some unreliability right after the upgrade when there is only one osd, so we will retry if needed
-	assert.Nil(s.T(), s.k8sh.ReadFromPodRetry("", podName, preFilename, message, 5))
-	assert.Nil(s.T(), s.k8sh.WriteToPodRetry("", podName, postFilename, message, 5))
-	assert.Nil(s.T(), s.k8sh.ReadFromPodRetry("", podName, postFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.ReadFromPodRetry("", podName, preFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.WriteToPodRetry("", podName, postFilename, message, 5))
+	assert.NoError(s.T(), s.k8sh.ReadFromPodRetry("", podName, postFilename, message, 5))
```

Comment:
Do we also need to add a check at the end of the test that the OSDs are all successfully running? If one OSD goes down after the upgrade, the test could still succeed, right? Separate PR?

Reply:
Unless I'm misinterpreting your thought, that is tested above starting with the line 
```go
osdsNotOldVersion = fmt.Sprintf("app=rook-ceph-osd,rook-version!=%s", oldRookVersion)
```

Reply:
Right, that verifies the pod specs have been specified. I'm asking an additional question to know if the OSDs are actually healthy after the upgrade. An OSD could get upgraded, the pod could be `running` for a brief time, then the OSD could crash. For example, are all the PGs clean?

Reply:
Oh I see. That is a whole bigger thing for sure. Checking that OSDs are healthy and stable (by checking healthy multiple times) is one thing we could do. Checking Ceph's health is another, but we can't do it reliably across different environments because we will have HEALTH_WARN on improper pg/pgp counts unless we can enable pg autoscaling I think. Or unless we compute some PG counts in the test which seems a bit much maybe.